File: app/src/main/java/org/mesonet/app/reflection/MesonetModelParser.java
package org.mesonet.app.reflection;


import java.lang.reflect.Field;

public class MesonetModelParser
{
    private static MesonetModelParser sModelParser;



    public static MesonetModelParser GetInstance()
    {
        if(sModelParser == null)
            sModelParser = new MesonetModelParser();

        return sModelParser;
    }



    private MesonetModelParser(){}



    public <T> T Parse(Class<T> inModelClass, String inStrValues)
    {
        if(inModelClass == null)
            return null;

        T model = null;
        try {
            model = inModelClass.newInstance();
        }
        catch (IllegalAccessException | InstantiationException e)
        {
            e.printStackTrace();
        }

        if(model == null || inStrValues == null)
            return model;

        String[] strFields = inStrValues.split(",");
        Field[] fields = inModelClass.getDeclaredFields();

        for(int i = 0; i < strFields.length; i++)
        {
            String[] strField = strFields[i].split("=");

            if(strField.length != 2)
                continue;

            String fieldName = strField[0];
            String fieldValue = strField[1];

            for(int j = 0; j < fields.length; j++)
            {
                if(fields[j].getName().equals(fieldName))
                {
                    fields[j].setAccessible(true);
                    if(fields[j].getType() == byte.class || fields[j].getType() == Byte.class)
                    {
                        try
                        {
                            fields[j].set(model, Byte.parseByte(fieldValue));
                        }
                        catch (IllegalAccessException | IllegalArgumentException e)
                        {
                            e.printStackTrace();
                        }
                    }
                    if(fields[j].getType() == short.class || fields[j].getType() == Short.class)
                    {
                        try
                        {
                            fields[j].set(model, Short.parseShort(fieldValue));
                        }
                        catch (IllegalAccessException | IllegalArgumentException e)
                        {
                            e.printStackTrace();
                        }
                    }
                    if(fields[j].getType() == int.class || fields[j].getType() == Integer.class)
                    {
                        try
                        {
                            fields[j].set(model, Integer.parseInt(fieldValue));
                        }
                        catch (IllegalAccessException | IllegalArgumentException e)
                        {
                            e.printStackTrace();
                        }
                    }
                    if(fields[j].getType() == long.class || fields[j].getType() == Long.class)
                    {
                        try
                        {
                            fields[j].set(model, Long.parseLong(fieldValue));
                        }
                        catch (IllegalAccessException | IllegalArgumentException e)
                        {
                            e.printStackTrace();
                        }
                    }
                    if(fields[j].getType() == float.class || fields[j].getType() == Float.class)
                    {
                        try
                        {
                            fields[j].set(model, Float.parseFloat(fieldValue));
                        }
                        catch (IllegalAccessException | IllegalArgumentException e)
                        {
                            e.printStackTrace();
                        }
                    }
                    if(fields[j].getType() == double.class || fields[j].getType() == Double.class)
                    {
                        try
                        {
                            fields[j].set(model, Double.parseDouble(fieldValue));
                        }
                        catch (IllegalAccessException | IllegalArgumentException e)
                        {
                            e.printStackTrace();
                        }
                    }
                    if(fields[j].getType() == char.class || fields[j].getType() == Character.class)
                    {
                        try
                        {
                            fields[j].set(model, fieldValue.charAt(0));
                        }
                        catch (IllegalAccessException | IllegalArgumentException e)
                        {
                            e.printStackTrace();
                        }
                    }
                    if(fields[j].getType() == boolean.class || fields[j].getType() == Boolean.class)
                    {
                        try
                        {
                            fields[j].set(model, Boolean.parseBoolean(fieldValue));
                        }
                        catch (IllegalAccessException | IllegalArgumentException e)
                        {
                            e.printStackTrace();
                        }
                    }
                    if(fields[j].getType() == Number.class)
                    {
                        try
                        {
                            fields[j].set(model, Double.parseDouble(fieldValue));
                        }
                        catch (IllegalAccessException | IllegalArgumentException e)
                        {
                            e.printStackTrace();
                        }
                    }
                    if(fields[j].getType() == String.class)
                    {
                        try
                        {
                            fields[j].set(model, fieldValue);
                        }
                        catch (IllegalAccessException | IllegalArgumentException e)
                        {
                            e.printStackTrace();
                        }
                    }

                    fields[i].setAccessible(false);
                }
            }
        }

        return model;
    }
}

File: app/src/main/java/org/mesonet/app/reflection/MesonetModelParser.kt
package org.mesonet.app.reflection


import java.lang.reflect.Field

import javax.inject.Inject
import kotlin.reflect.KClass
import kotlin.reflect.KMutableProperty
import kotlin.reflect.full.declaredMemberProperties
import kotlin.reflect.full.primaryConstructor
import kotlin.reflect.full.starProjectedType
import kotlin.reflect.full.withNullability
import kotlin.reflect.jvm.isAccessible

class MesonetModelParser @Inject
internal constructor() {


    fun <T : Any> Parse(inModelClass: KClass<T>?, inStrValues: String?): T? {
        if (inModelClass == null)
            return null

        var model: T? = null
        try {
            model = inModelClass.primaryConstructor?.call()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: InstantiationException) {
            e.printStackTrace()
        }

        if (model == null || inStrValues == null)
            return model

        val strFields = inStrValues.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val fields = ArrayList(inModelClass.declaredMemberProperties)

        for (i in strFields.indices) {
            val strField = strFields[i].split("=".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

            if (strField.size != 2)
                continue

            val fieldName = strField[0]
            val fieldValue = strField[1]

            for (j in fields.indices) {
                if (fields[j].name == fieldName) {
                    fields[j].isAccessible = true
                    if (fields[j].returnType == Byte::class.javaPrimitiveType ||
                        fields[j].returnType == Byte::class.javaObjectType ||
                        fields[j].returnType == Byte::class ||
                        fields[j].returnType == Byte::class.starProjectedType.withNullability(true))
                    {
                        try {
                            (fields[j] as KMutableProperty<*>).setter.call(model, java.lang.Byte.parseByte(fieldValue))
                        } catch (e: IllegalAccessException) {
                            e.printStackTrace()
                        } catch (e: IllegalArgumentException) {
                            e.printStackTrace()
                        }

                    }
                    if (fields[j].returnType == Short::class.javaPrimitiveType ||
                            fields[j].returnType == Short::class.javaObjectType ||
                            fields[j].returnType == Short::class ||
                            fields[j].returnType == Short::class.starProjectedType.withNullability(true)) {
                        try {
                            (fields[j] as KMutableProperty<*>).setter.call(model, java.lang.Short.parseShort(fieldValue))
                        } catch (e: IllegalAccessException) {
                            e.printStackTrace()
                        } catch (e: IllegalArgumentException) {
                            e.printStackTrace()
                        }

                    }
                    if (fields[j].returnType == Int::class.javaPrimitiveType ||
                            fields[j].returnType == Int::class.javaObjectType ||
                            fields[j].returnType == Int::class ||
                            fields[j].returnType == Int::class.starProjectedType.withNullability(true)) {
                        try {
                            (fields[j] as KMutableProperty<*>).setter.call(model, Integer.parseInt(fieldValue))
                        } catch (e: IllegalAccessException) {
                            e.printStackTrace()
                        } catch (e: IllegalArgumentException) {
                            e.printStackTrace()
                        }

                    }
                    if (fields[j].returnType == Long::class.javaPrimitiveType ||
                            fields[j].returnType == Long::class.javaObjectType ||
                            fields[j].returnType == Long::class ||
                            fields[j].returnType == Long::class.starProjectedType.withNullability(true)) {
                        try {
                            (fields[j] as KMutableProperty<*>).setter.call(model, java.lang.Long.parseLong(fieldValue))
                        } catch (e: IllegalAccessException) {
                            e.printStackTrace()
                        } catch (e: IllegalArgumentException) {
                            e.printStackTrace()
                        }

                    }
                    if (fields[j].returnType == Float::class.javaPrimitiveType ||
                            fields[j].returnType == Float::class.javaObjectType ||
                            fields[j].returnType == Float::class ||
                            fields[j].returnType == Float::class.starProjectedType.withNullability(true)) {
                        try {
                            (fields[j] as KMutableProperty<*>).setter.call(model, java.lang.Float.parseFloat(fieldValue))
                        } catch (e: IllegalAccessException) {
                            e.printStackTrace()
                        } catch (e: IllegalArgumentException) {
                            e.printStackTrace()
                        }

                    }
                    if (fields[j].returnType == Double::class.javaPrimitiveType ||
                            fields[j].returnType == Double::class.javaObjectType ||
                            fields[j].returnType == Double::class ||
                            fields[j].returnType == Double::class.starProjectedType.withNullability(true)) {
                        try {
                            (fields[j] as KMutableProperty<*>).setter.call(model, java.lang.Double.parseDouble(fieldValue))
                        } catch (e: IllegalAccessException) {
                            e.printStackTrace()
                        } catch (e: IllegalArgumentException) {
                            e.printStackTrace()
                        }

                    }
                    if (fields[j].returnType == Char::class.javaPrimitiveType ||
                            fields[j].returnType == Char::class.javaObjectType ||
                            fields[j].returnType == Char::class ||
                            fields[j].returnType == Char::class.starProjectedType.withNullability(true)) {
                        try {
                            (fields[j] as KMutableProperty<*>).setter.call(model, fieldValue[0])
                        } catch (e: IllegalAccessException) {
                            e.printStackTrace()
                        } catch (e: IllegalArgumentException) {
                            e.printStackTrace()
                        }

                    }
                    if (fields[j].returnType == Boolean::class.javaPrimitiveType ||
                            fields[j].returnType == Boolean::class.javaObjectType ||
                            fields[j].returnType == Boolean::class ||
                            fields[j].returnType == Boolean::class.starProjectedType.withNullability(true)) {
                        try {
                            (fields[j] as KMutableProperty<*>).setter.call(model, java.lang.Boolean.parseBoolean(fieldValue))
                        } catch (e: IllegalAccessException) {
                            e.printStackTrace()
                        } catch (e: IllegalArgumentException) {
                            e.printStackTrace()
                        }

                    }
                    if (fields[j].returnType == Number::class.javaPrimitiveType ||
                            fields[j].returnType == Number::class.javaObjectType ||
                            fields[j].returnType == Number::class ||
                            fields[j].returnType == Number::class.starProjectedType.withNullability(true)) {
                        try {
                            (fields[j] as KMutableProperty<*>).setter.call(model, java.lang.Double.parseDouble(fieldValue))
                        } catch (e: IllegalAccessException) {
                            e.printStackTrace()
                        } catch (e: IllegalArgumentException) {
                            e.printStackTrace()
                        }

                    }
                    if (fields[j].returnType == String::class.javaPrimitiveType ||
                            fields[j].returnType == String::class.javaObjectType ||
                            fields[j].returnType == String::class ||
                            fields[j].returnType == String::class.starProjectedType.withNullability(true)) {
                        try {
                            (fields[j] as KMutableProperty<*>).setter.call(model, fieldValue)
                        } catch (e: IllegalAccessException) {
                            e.printStackTrace()
                        } catch (e: IllegalArgumentException) {
                            e.printStackTrace()
                        }

                    }

                    fields[i].isAccessible = false
                }
            }
        }

        return model
    }
}
